packages/eslint-plugin/src/rules/lines-around-comment.ts

```diff
@@ -0,0 +1,528 @@
+ */
+function getEmptyLineNums(lines: string[]): number[] {
+  const emptyLines = lines
+    .map((line, i) => ({
+      code: line.trim(),
+      num: i + 1,
+    }))
+    .filter(line => !line.code)
+    .map(line => line.num);
+
+  return emptyLines;
+}
+
+/**
+ * Return an array with with any line numbers that contain comments.
+ */
+function getCommentLineNums(comments: TSESTree.Comment[]): number[] {
+  const lines: number[] = [];
+
+  comments.forEach(token => {
+    const start = token.loc.start.line;
+    const end = token.loc.end.line;
+
+    lines.push(start, end);
+  });
+  return lines;
+}
+
+export default util.createRule<Options, MessageIds>({
+  name: 'lines-around-comment',
+  meta: {
+    type: 'layout',
+    docs: {
+      description: 'Require empty lines around comments',
+      recommended: false,
+      extendsBaseRule: true,
+    },
+    schema: {
+      type: 'array',
+      items: [
+        {
+          type: 'object',
+          properties: {
+            beforeBlockComment: {
+              type: 'boolean',
+              default: true,
+            },
+            afterBlockComment: {
+              type: 'boolean',
+              default: false,
+            },
+            beforeLineComment: {
+              type: 'boolean',
+              default: false,
+            },
+            afterLineComment: {
+              type: 'boolean',
+              default: false,
+            },
+            allowBlockStart: {
+              type: 'boolean',
+              default: false,
+            },
+            allowBlockEnd: {
+              type: 'boolean',
+              default: false,
+            },
+            allowClassStart: {
+              type: 'boolean',
+            },
+            allowClassEnd: {
+              type: 'boolean',
+            },
+            allowObjectStart: {
+              type: 'boolean',
+            },
+            allowObjectEnd: {
+              type: 'boolean',
+            },
+            allowArrayStart: {
+              type: 'boolean',
+            },
+            allowArrayEnd: {
+              type: 'boolean',
+            },
+            allowInterfaceStart: {
+              type: 'boolean',
+            },
+            allowInterfaceEnd: {
+              type: 'boolean',
+            },
+            allowTypeStart: {
+              type: 'boolean',
+            },
+            allowTypeEnd: {
+              type: 'boolean',
+            },
+            ignorePattern: {
+              type: 'string',
+            },
+            applyDefaultIgnorePatterns: {
+              type: 'boolean',
+            },
+          },
+          additionalProperties: false,
+        },
+      ],
+    },
+    fixable: baseRule.meta.fixable,
+    hasSuggestions: baseRule.meta.hasSuggestions,
+    messages: baseRule.meta.messages,
+  },
+  defaultOptions: [
+    {
+      beforeBlockComment: true,
+    },
+  ],
+  create(context, [_options]) {
+    const options = _options!;
```

Comment:
Yeah, this kind of thing is unfortunate. Filed #5439 to see if we can clean it up.

(not blocking for this PR, the `!` should be linted against once it's unnecessary)